File: Java/Queue.java
import java.util.Arrays;
import java.util.NoSuchElementException;

/**
 * Created by Chris Medina on 5/14/2014.
 * Purpose: Array-based implementation of a Queue
 */
public class Queue<E> {
    private int size;
    private int max_size;
    private E[] array;

    public Queue (int max) {
        this.max_size = max;
        this.size = 0;
        this.array = (E[]) new Object[max_size];
    }

    public Queue () {
        this(10);
    }

    /**
     * Determine if this Queue is empty
     *
     * @return true, if empty
     */
    public boolean isEmpty () {
        return (size == 0);
    }

    /**
     * Determine if this Queue is full
     *
     * @return true, if full
     */
    public boolean isFull () {
        return (size == max_size);
    }

    private void ensureCapacity () {
        if (isFull()) {
            max_size = max_size * 3;
            array = Arrays.copyOf(array, max_size);
        }
    }

    /**
     * Insert into this queue on a max-value priority basis, such that larger numbers appear at the front of the array
     *
     * @param val the value to insert
     */
    public void priority_insert (E val) {
        ensureCapacity();
        if (isEmpty()) {
            array[0] = val;
        } else {
            int j = size - 1;
            while (j >= 0 && ((Comparable) val).compareTo(array[j]) > 0) {
                array[j + 1] = array[j];
                j--;
            }
            array[j + 1] = val;
        }
        size++;
    }

    /**
     * Insert this value into the rear of the array
     *
     * @param val the value to insert
     */
    public void insert (E val) {
        ensureCapacity();
        array[size++] = val;
    }

    /**
     * Remove this value from the front of the Queue
     *
     * @return the value removed
     */
    public E remove () {
        if (isEmpty())
            throw new NoSuchElementException("Queue is empty");
        else {
            E temp = array[0];
            size--;
            System.arraycopy(array, 1, array, 0, size);
            return temp;
        }
    }

    /**
     * Displays the next value to be removed
     *
     * @return the first value in the array (next to be removed)
     */
    public E peek () {
        if (isEmpty())
            throw new NoSuchElementException("Queue is empty");
        return array[0];
    }

    public int size () {
        return size;
    }

    public String toString () {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < size; i++) {
            if (i < size - 1)
                sb.append(array[i] + ", ");
            else
                sb.append(array[i]);
        }
        sb.append("]");
        return sb.toString();
    }

}

File: C-Sharp/Queue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OOP
{
    /// <summary>
    /// Date: 5-25-2014
    /// Represents a generic Queue data structure. Items are added to the end and removed from the front
    /// </summary>
    /// <typeparam name="E">Generic, type-safe implementation</typeparam>
    public class Queue<E>
    {
        private E[] array;
        private int max_size;
        
        public int size { get; private set; }

        public Queue (int max) {
            this.max_size = max;
            this.array = new E[max];
            this.size = 0;
        }

        public Queue ()
            : this(10) {

        }

        /// <summary>
        /// Determine if this Queue is empty
        /// </summary>
        /// <returns>true, if it has no items</returns>
        public bool isEmpty () {
            return (size == 0);
        }

        /// <summary>
        /// Determine if this Queue is full
        /// </summary>
        /// <returns>true, if is full. Once reached, the array dynamically grows</returns>
        public bool isFull () {
            return (size == max_size);
        }

        private void ensureCapacity () {
            if (isFull()) {
                max_size *= 3;
                Array.Resize(ref array, max_size);
            }
        }

        /// <summary>
        /// Append this item to the end of the queue
        /// </summary>
        /// <param name="item">the item to append</param>
        public void enqueue (E item) {
            ensureCapacity();
            array[size++] = item;
        }

        /// <summary>
        /// Remove the item at the front of the queue, or throw exception if empty
        /// </summary>
        /// <returns>the item removed from the front of queue</returns>
        public E dequeue () {
            if (isEmpty())
                throw new InvalidOperationException("Queue is empty");
            E temp = array[0];

            for (int i = 0; i < size - 1; i++) {
                array[i] = array[i + 1];
            }

            Array.Resize(ref array, array.Length - 1);

            size--;
            return temp;
        }

        /// <summary>
        /// Show the item that will be deleted next 
        /// </summary>
        /// <returns>the item to be removed next upon calling dequeue</returns>
        public E peek () {
            if (isEmpty())
                throw new InvalidOperationException("Queue is empty");
            return array[0];
        }

        public override string ToString () {
            StringBuilder sb = new StringBuilder("[");

            for (int i = 0; i < size; i++) {
                if (i < size - 1) {
                    sb.Append(array[i] + ", ");
                }
                else {
                    sb.Append(array[i]);
                }
            }

            sb.Append("]");
            return sb.ToString();
        }

    }

}
